flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java

```diff
@@ -0,0 +1,146 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.nodes.resource.batch.parallelism;
+
+import org.apache.flink.table.plan.nodes.exec.ExecNode;
+import org.apache.flink.table.plan.nodes.physical.batch.BatchExecExchange;
+import org.apache.flink.table.plan.nodes.physical.batch.BatchExecSink;
+import org.apache.flink.table.plan.nodes.physical.batch.BatchExecUnion;
+
+import org.apache.calcite.rel.RelDistribution;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Build shuffleStages.
+ */
+public class ShuffleStageGenerator {
+
+	private final Map<ExecNode<?, ?>, ShuffleStage> nodeShuffleStageMap = new LinkedHashMap<>();
+	private final Map<ExecNode<?, ?>, Integer> nodeToFinalParallelismMap;
+
+	private ShuffleStageGenerator(Map<ExecNode<?, ?>, Integer> nodeToFinalParallelismMap) {
+		this.nodeToFinalParallelismMap = nodeToFinalParallelismMap;
+	}
+
+	public static Map<ExecNode<?, ?>, ShuffleStage> generate(List<ExecNode<?, ?>> sinkNodes, Map<ExecNode<?, ?>, Integer> finalParallelismNodeMap) {
+		ShuffleStageGenerator generator = new ShuffleStageGenerator(finalParallelismNodeMap);
+		sinkNodes.forEach(generator::buildShuffleStages);
+		Map<ExecNode<?, ?>, ShuffleStage> result = generator.getNodeShuffleStageMap();
+		result.values().forEach(s -> {
+			List<ExecNode<?, ?>> virtualNodeList = s.getExecNodeSet().stream().filter(ShuffleStageGenerator::isVirtualNode).collect(toList());
+			virtualNodeList.forEach(s::removeNode);
+		});
+		return generator.getNodeShuffleStageMap().entrySet().stream()
+				.filter(x -> !isVirtualNode(x.getKey()))
+				.collect(Collectors.toMap(Map.Entry::getKey,
+						Map.Entry::getValue,
+						(e1, e2) -> e1,
+						LinkedHashMap::new));
+	}
+
+	private void buildShuffleStages(ExecNode<?, ?> execNode) {
+		if (nodeShuffleStageMap.containsKey(execNode)) {
+			return;
+		}
+		for (ExecNode<?, ?> input : execNode.getInputNodes()) {
+			buildShuffleStages((input));
+		}
+
+		if (execNode.getInputNodes().isEmpty()) {
+			// source node
+			ShuffleStage shuffleStage = new ShuffleStage();
+			shuffleStage.addNode(execNode);
+			if (nodeToFinalParallelismMap.containsKey(execNode)) {
+				shuffleStage.setParallelism(nodeToFinalParallelismMap.get(execNode), true);
+			}
+			nodeShuffleStageMap.put(execNode, shuffleStage);
+		} else if (execNode instanceof BatchExecExchange && !(((BatchExecExchange) execNode).getDistribution().getType() == RelDistribution.Type.RANGE_DISTRIBUTED)) {
+			// do nothing
+		} else if (execNode instanceof BatchExecSink) {
```

Comment:
Here you consider `BatchExecSink`, and you exclude `BatchExecSink` before?

Reply:
> Here you consider `BatchExecSink`, and you exclude `BatchExecSink` before?

Double exclude, I fix it.